perfmetrics/scripts/bigquery/experiments_gcsfuse_bq.py

```diff
@@ -0,0 +1,348 @@
+    Returns:
+      google.cloud.bigquery.table.Table: The table in BigQuery
+    """
+    table_ref = self.dataset_ref.table(table_id)
+    table = self.client.get_table(table_ref)
+    return table
+
+  def _execute_query(self, query) -> QueryJob:
+    """Executes the query in BigQuery and raises an exception if query
+       execution could not be completed.
+
+    Args:
+      query (str): Query that will be executed in BigQuery.
+
+    Raises:
+      Exception: If query execution failed.
+    """
+    job = self.client.query(query)
+    if job.errors:
+      for error in job.errors:
+        raise Exception(f"Error message: {error['message']}")
+    return job
+
+  def _check_if_config_valid(self, exp_config_id) -> bool:
+    """Checks if exp_config_id exists in the experiment_configuration table.
+
+    Args:
+      exp_config_id (str): An id that uniquely identifies an experiment
+
+    Returns:
+      bool: Returns true if exp_config_id exists, false otherwise.
+    """
+    query_check_if_config_valid = """
+      SELECT *
+      FROM `{}.{}.{}`
+      WHERE configuration_id = '{}'
+    """.format(self.project_id, self.dataset_id, constants.CONFIGURATION_TABLE_ID, exp_config_id)
+
+    job = self._execute_query(query_check_if_config_valid)
+    row_count = job.result().total_rows
+    if row_count:
+      return True
+    return False
+
+  def _delete_rows_incomplete_transaction(self, table_id = None, config_id = None, start_time_build = None):
+    """Helper function for _insert_row. If insertion of some nth row fails,
+    this method deletes (n-1) rows that were inserted before
+
+    Args:
+      table_id (str): ID of table to which results are being uploaded
+      config_id (str): config_id of the experiment for which results are being uploaded
+      start_time_build (timestamp): Start epoch time of the build
+    """
+    if config_id:
+      query_delete_if_row_exists = """
+        DELETE FROM `{}.{}.{}`
+        WHERE configuration_id = '{}'
+        AND start_time_build = '{}'
+      """.format(self.project_id, self.dataset_id, table_id, config_id, start_time_build)
+      job = self._execute_query(query_delete_if_row_exists)
+
+  def _insert_rows(self, table, rows_to_insert, table_id = None, config_id = None, start_time_build = None):
+    """Insert rows in table. If insertion of some nth row fails, delete (n-1) rows
+    that were inserted before and raise an exception
+
+    Args:
+      table (str): Table in which rows are being inserted
+      rows_to_insert (str): Rows to insert in the table
+      table_id (str): ID of table to which results are being uploaded
+      config_id (str): config_id of the experiment for which results are being uploaded
+      start_time_build (timestamp): Start epoch time of the build
+
+    Raises:
+      Exception: If some row insertion failed.
+    """
+    try:
+      result = self.client.insert_rows(table, rows_to_insert)
+      if result:
+        raise Exception(f'{result}')
```

Comment:
Should we also delete rows if insert_rows function insert some rows, doesn't fail and contain some errors in result ?

Reply:
@RuchikaSharma-8 corrected me it is deleting in both cases.